File: restring/src/main/java/com/ice/restring/RestringResources.java
package com.ice.restring;

import android.content.res.Resources;
import android.os.Build;
import android.support.annotation.NonNull;
import android.text.Html;

/**
 * This is the wrapped resources which will be provided by Restring.
 * For getting strings and texts, it checks the strings repository first and if there's a new string
 * that will be returned, otherwise it will fallback to the original resource strings.
 */
class RestringResources extends Resources {

    private final StringRepository stringRepository;

    RestringResources(@NonNull final Resources res,
                      @NonNull final StringRepository stringRepository) {
        super(res.getAssets(), res.getDisplayMetrics(), res.getConfiguration());
        this.stringRepository = stringRepository;
    }

    @NonNull
    @Override
    public String getString(int id) throws NotFoundException {
        String value = getStringFromRepository(id);
        if (value != null) {
            return value;
        }
        return super.getString(id);
    }

    @NonNull
    @Override
    public String getString(int id, Object... formatArgs) throws NotFoundException {
        String value = getStringFromRepository(id);
        if (value != null) {
            return String.format(value, formatArgs);
        }
        return super.getString(id, formatArgs);
    }

    @NonNull
    @Override
    public CharSequence getText(int id) throws NotFoundException {
        String value = getStringFromRepository(id);
        if (value != null) {
            return fromHtml(value);
        }
        return super.getText(id);
    }

    @Override
    public CharSequence getText(int id, CharSequence def) {
        String value = getStringFromRepository(id);
        if (value != null) {
            return fromHtml(value);
        }
        return super.getText(id, def);
    }

    private String getStringFromRepository(int id) {
        try {
            String stringKey = getResourceEntryName(id);
            return stringRepository.getString(RestringUtil.getCurrentLanguage(), stringKey);
        } catch (NotFoundException ex) {
            return null;
        }
    }

    private CharSequence fromHtml(String source) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
            //noinspection deprecation
            return Html.fromHtml(source);
        } else {
            return Html.fromHtml(source, Html.FROM_HTML_MODE_COMPACT);
        }
    }
}

File: restring/src/main/java/com/b3nedikt/restring/RestringResources.kt
package com.b3nedikt.restring

import android.content.res.Resources
import android.os.Build
import androidx.core.text.HtmlCompat


/**
 * This is the wrapped resources which will be provided by Restring.
 * For getting strings and texts, it checks the strings repository first and if there's a new string
 * that will be returned, otherwise it will fallback to the original resource strings.
 */
@Suppress("DEPRECATION")
internal class RestringResources(val res: Resources,
                                 private val stringRepository: StringRepository)
    : Resources(res.assets, res.displayMetrics, res.configuration) {

    @Throws(NotFoundException::class)
    override fun getString(id: Int): String {
        setLocale()

        val value = getStringFromRepository(id)
        return value ?: super.getString(id)
    }

    @Throws(NotFoundException::class)
    override fun getString(id: Int, vararg formatArgs: Any): String {
        setLocale()

        val value = getStringFromRepository(id)
        if (value != null) return String.format(value, *formatArgs)
        return super.getString(id, *formatArgs)
    }

    @Throws(NotFoundException::class)
    override fun getText(id: Int): CharSequence {
        setLocale()

        val value = getStringFromRepository(id)
        return value?.let { fromHtml(it) } ?: super.getText(id)
    }

    override fun getText(id: Int, def: CharSequence): CharSequence {
        setLocale()

        val value = getStringFromRepository(id)
        return value?.let { fromHtml(it) } ?: super.getText(id, def)
    }

    private fun getStringFromRepository(id: Int): String? {
        val currentLocale = RestringLocale.currentLocale
        val supportedLocales = stringRepository.supportedLocales

        val resultLocale = if (supportedLocales.contains(currentLocale)) {
            currentLocale
        } else {
            supportedLocales.find { it.language == currentLocale.language }
        }

        resultLocale ?: return null

        return try {
            val stringKey = getResourceEntryName(id)
            return stringRepository.getString(resultLocale, stringKey)
        } catch (e: NotFoundException) {
            null
        }
    }

    private fun setLocale() {
        if (RestringLocale.isInitial) return

        val conf = res.configuration

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            conf.setLocale(RestringLocale.currentLocale)
        } else {
            conf.locale = RestringLocale.currentLocale
        }

        res.updateConfiguration(conf, res.displayMetrics)
    }

    private fun fromHtml(source: String) =
            HtmlCompat.fromHtml(source, HtmlCompat.FROM_HTML_MODE_COMPACT)
}
